### ISSUE TITLE
Retrieval quality: bge-m3 sparse vectors are discarded and there is no cross-encoder rerank

### ISSUE THREAD

**Issue by Ryouku** [atvirokodosprendimai/agentsmemory#14]
agentsmemory embeds with **bge-m3** but uses only its dense output, and never reranks with a cross-encoder. Both halves of the bge-m3 design are on the table for free-ish, and neither is wired.

## What already exists (so this is an addition, not a rewrite)

Search today is genuinely hybrid on every backend:

- over-fetch `limit*3` vector neighbours — `internal/palace/service.go:413`
- Okapi-BM25 over the candidate text, corpus-relative IDF — `internal/palace/rank.go:63`
- fused `0.6*cosine + 0.4*bm25 + closetBoost` — `internal/palace/rank.go:161`

That is a solid lexical reranker and it should stay.

## Gap 1 — bge-m3's sparse and ColBERT vectors are discarded

Verified against a local Ollama with `bge-m3:latest`:

```
$ curl -s localhost:11434/api/embed -d '{"model":"bge-m3","input":"test"}' | jq 'keys'
["embeddings","load_duration","model","prompt_eval_count","total_duration"]
# 1 vector, 1024 dims — dense only
```

`internal/embed/ollama/ollama.go` is therefore correct for what Ollama offers; the limitation is upstream. bge-m3's multi-vector retrieval (dense + learned sparse + ColBERT) is the model's whole point, and the in-process BM25 is a hand-rolled stand-in for the sparse half — a good one, but corpus-relative IDF over ≤10 000 candidates is not a learned sparse index over the whole collection.

Getting sparse out needs **TEI** (`text-embeddings-inference`) or FlagEmbedding, and Qdrant named/sparse vectors with server-side fusion to store and query it.

## Gap 2 — no cross-encoder rerank

`bge-m3` and `bge-reranker-v2-m3` are a matched pair; only the first is wired. A cross-encoder over the top ~50 fused candidates is the single largest precision win available here, and it is the piece that makes `am_search` results worth their token cost in an agent's context window.

Ollama cannot serve it (`/api/embed` is embeddings-only); TEI's `/rerank` endpoint can.

## Suggested shape

- A `Reranker` interface defined at the consumer in `internal/palace`, with a no-op default so nothing changes when it is unconfigured (`RERANK_URL` empty).
- `rankHybrid` keeps its job — it produces the candidate pool. The cross-encoder reorders the top-K of that pool, and its score replaces `Fused` for the returned page.
- Config: `RERANK_URL`, `RERANK_MODEL`, `RERANK_TOP_K` (default ~50), all optional.
- Sparse support goes behind the embedder seam: an embedder that returns `(dense, sparse)` and a Qdrant driver that upserts named vectors. `sync` replays the dense half from SQLite without re-embedding; the sparse half needs a re-embed pass, which the existing `embedded_at` queue shape can drive.

See the companion issue on the maxed-out compose profile for how these get shipped without breaking the "one container, no deps" default.

**Comment by Ryouku** [comment]
Closing with a split verdict, because half of this turned out not to be ours to fix.

**Cross-encoder rerank: shipped.** A `bge-reranker-v2-m3` pass runs behind `RERANK_*` config, with the service in `docker-compose.full.yml`, and the eval harness carries reranked arms so the gain is measurable rather than asserted.

**bge-m3 sparse weights: blocked upstream, and now recorded so nobody re-derives it.** The sparse head is unreachable through TEI. `/embed_sparse` requires SPLADE pooling — sparse weights read off an MLM head as logits over the vocabulary — whereas bge-m3 produces its weights from a separate trained `sparse_linear` layer TEI does not load. Probed against a live `BAAI/bge-m3` on TEI 1.9.3; the finding and the upstream PR that would add `--pooling m3_sparse` are written into `internal/embed/teiembed/teiembed.go:15`.

The lexical role that sparse was wanted for is served instead by BM25 in the hybrid ranker, which is swept in the eval harness. So the retrieval-quality goal is met by a different route; reopen if the upstream pooling flag lands and someone wants the true sparse head.
